src/cryptonote_core/tx_pool.cpp

```diff
@@ -1626,13 +1711,94 @@ namespace cryptonote
     return n_removed;
   }
   //---------------------------------------------------------------------------------
+  void tx_memory_pool::add_tx_to_transient_lists(const crypto::hash& txid, double fee, time_t receive_time)
+  {
+    m_txs_by_fee_and_receive_time.emplace(std::pair<double, time_t>(fee, receive_time), txid);
+    time_t now = time(NULL);
+    m_added_txs_by_time.insert(std::make_pair(now, txid));
```

Comment:
I encountered a one-off case I haven't been able to repro yet where it *didn't* re-insert the tx into the transient list upon broadcast, and in that case, even with the above proposed approaches, there would still be a vector to determine sensitive tx's (and I believe there would be a bug in that case: seems it would be possible for incremental polling to miss the tx was added to the pool). Still investigating this

Reply:
> I encountered a one-off case I haven't been able to repro yet 

Alright, I wait what result your investigation will bring.

> Alternative approach: you could keep track of sensitivity with a boolean here like you do for removed, allow duplicate hashes in m_added_txs_by_time, and when removing later, iterate over every single m_added_txs_by_time and remove matching txid's. Aka remove the break. Probably safer this way, but less performant.

I do already allow duplicates hashes in `m_added_txs_by_time`, don't I? And always scanning over the whole array instead of breaking as soon as I find the first entry with correct id to remove would be trivial, and I ask what an additional sensitivity boolean would help or why it would be even needed.

And concerning perfomance I ask myself anyway whether we worry too much ...

Reply:
Status update on the investigation...

I can repro the case where duplicates aren't re-inserted into the transient list. When submitting txs over the clearnet, and my node is in the "fluff" epoch of Dandelion++ (at any given time, a node has a [1/5 chance](https://github.com/monero-project/monero/blob/9a124f681119855949f6406ecd69c2ad91da9770/src/cryptonote_protocol/levin_notify.cpp#L701) of being in the "fluff" epoch), the node will by default set the pool tx to the "fluff" state without waiting to hear back from another node after relaying [thanks to this](https://github.com/monero-project/monero/blob/9a124f681119855949f6406ecd69c2ad91da9770/src/cryptonote_protocol/levin_notify.cpp#L573).

Then, because the tx is in the pool in the "fluff" state already, once it hears back from another node that the tx has been broadcasted successfully, it won't attempt to re-add the tx into the pool because [this is true](https://github.com/monero-project/monero/blob/67e5ca9ad6f1c861ad315476a88f9d36c38a0abb/src/cryptonote_core/cryptonote_core.cpp#L1036) ([this comes back as true](https://github.com/monero-project/monero/blob/67e5ca9ad6f1c861ad315476a88f9d36c38a0abb/src/blockchain_db/blockchain_db.cpp#L70-L71) for `relay_category::legacy`). Thus, it doesn't re-insert into the pool/transient list in this case, unlike the other cases I tested (clearnet + stem epoch, tor + fluff epoch, tor + stem epoch all inserted duplicates into the transient list).

> I do already allow duplicates hashes in m_added_txs_by_time, don't I?

Yep, I was saying keeping it like that still seemed like an option along with those other changes

> And I ask what an additional sensitivity boolean would help or why it would be even needed.

A sensitivity boolean here isn't the only way to accomplish the goal (of avoiding revealing sensitive txs over the incremental RPC endpoint), but the point is that when returning the added pool tx's over the incremental RPC endpoint, we need to make sure not to reveal that a tx *was* sensitive in the past, even if a client includes a `pool_info_since` earlier than the first duplicate. Revealing a tx *was* sensitive seems close-to-but-not-quite-as-bad as revealing a tx is sensitive at the time that it is sensitive. Returning a duplicate reveals a tx *was* sensitive, so there needs to be a way to avoid doing that in all cases. Adding a `sensitive` boolean here seems like one way (only return the non-sensitive entries), but considering this case of the fluff epoch above, it seems pretty challenging to do safely.

Probably easier (and probably what you were thinking I'd guess): right in [this spot](https://github.com/monero-project/monero/pull/8076/files#r865601662), keep track of inserted tx's being sure not to return duplicates. Ideally the txs are also returned sorted by `meta.receive_time`.

It's difficult to reason through potential races here, being completely honest.

Reply:
Sorry for the double comment -- I believe the PR as is and the solution I described above is also vulnerable to the following race condition when a clearnet daemon is in the "fluff" epoch:

1. Client submits tx
2. At time t0, daemon adds it to pool with `relay_method::local` and inserts into `m_added_txs_by_time`
3. At time t1, client requests pool info from before time t0; daemon doesn't respond with tx because it's sensitive
4. At time t2, daemon updates pool tx to no longer be sensitive
5. Client requests pool info since t1; daemon does NOT include tx, since it was added before t1

Thus, the daemon would never return the tx via the incremental endpoint. I believe this would reveal definitively that the tx was submitted through the daemon if it can be triggered successfully, because only a tx submitted through the node while in the "fluff" epoch appears to have this potential for being unquery-able via the incremental endpoint. It's also technically possible for a normal user to never see that their tx broadcasts successfully because of this until the tx is included in a block.

I believe a solution for this is to make sure that once the tx reaches the broadcasted state in the pool, at that point it definitely gets (re)-inserted into `m_added_txs_by_time` with a later time. Grep for `update_txpool_tx` to see more code paths where this can happen.

Reply:
I confess that unfortunately this goes over my head now, mostly because I don't know much about Dandelion++ and its implementation in the Monero code base. I think I am unable to code the proper solution all on my own, and you will have to closely guide me to a solution in code.

I think I basically understand what you say here:

> I believe a solution for this is to make sure that once the tx reaches the broadcasted state in the pool, at that point it definitely gets (re)-inserted into m_added_txs_by_time with a later time. Grep for update_txpool_tx to see more code paths where this can happen.

You talk of a move, right? The old tx with its old time (which was sensitive when it got inserted) gets deleted, and re-inserted with the time of the call to `update_pool_tx`?

And do you think that if this is done properly it's the definite solution to all the problems that you unearthed here regarding Dandelion++ versus cache, leaks, race conditions and everything? Or in other words, if we ship around this cliff as well, the PR is ready?

If yes, which state transition or state transitions done with the help of the `update_txpool_tx` exactly will trigger the move / re-insert?

Or maybe it's as simple as making sure there are no duplicates and always only keep the most recent entry for the same tx id? Because we don't see any unproblematic cases with the tx present twice in the map?

Reply:
Alright, think I've got my head around the ideal path forward. A guide on further changes:

1. I think a "move" as you describe above is the way to go. `add_tx_to_transient_lists` could just check if the tx exists already first, then "move" it to the new `receive_time` if so. Less efficient, but significantly easier to reason that it is safe than when allowing duplicates case, which has more surface area to think through and deal with properly.

2. When a tx is upgrading from non-sensitive to sensitive in [`set_relayed`](https://github.com/monero-project/monero/blob/9a124f681119855949f6406ecd69c2ad91da9770/src/cryptonote_core/tx_pool.cpp#L823-L862), update its `receive_time` and call `add_tx_to_transient_list`
	- See [this snippet](https://paste.debian.net/1241229/) for a sample implementation
	- This ensures that there will be an entry at the time the tx is thought to have broadcast successfully, preventing the [above circumstance](https://github.com/monero-project/monero/pull/8076#discussion_r869061356).

> And do you think that if this is done properly it's the definite solution to all the problems that you unearthed here regarding Dandelion++ versus cache, leaks, race conditions and everything? Or in other words, if we ship around this cliff as well, the PR is ready?

I haven't been able to identify any related issues with Dandelion++ yet with certainty and have spent a solid chunk of time digging; whereas with this code as is, I can and do think they are solve-able :)

Bonus 1 (possible future PR): Include something like a `broadcasted` boolean in the response over the incremental endpoint. Otherwise, a client that is able to see sensitive txs would confusingly be able to see the same tx twice when including a `pool_info_since` at t0, and then again at t1 if it broadcasts successfully after t0. The `broadcasted` boolean provides the client a reason why it would see the same tx twice, even if sending `pool_info_since` at t0 and then again at t1.

Bonus 2 (possible future PR): I would also advocate for a new `unconfirmed_transfer_details::m_state` that can differentiate between "in pool and broadcasted" and "in pool and unbroadcasted"